via/db.py

```diff
@@ -0,0 +1,57 @@
+import alembic.command
+import alembic.config
+import zope.sqlalchemy
+from sqlalchemy import MetaData, create_engine, text
+from sqlalchemy.exc import ProgrammingError
+from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Session
+
+
+class Base(MappedAsDataclass, DeclarativeBase):
+    metadata = MetaData(
+        naming_convention={
+            "ix": "ix_%(column_0_N_label)s",
+            "uq": "uq_%(table_name)s_%(column_0_N_name)s",
+            "ck": "ck_%(table_name)s_%(constraint_name)s",
+            "fk": "fk_%(table_name)s_%(column_0_N_name)s_%(referred_table_name)s_%(referred_column_0_N_name)s",
+            "pk": "pk_%(table_name)s",
+        }
+    )
+
+
+def stamp(engine):  # pragma: no cover
+    """If the DB isn't already stamped then stamp it with the latest revision."""
+    stamped = False
+
+    with engine.connect() as connection:
+        try:
+            if connection.execute(text("select * from alembic_version")).first():
+                stamped = True
+        except ProgrammingError:
+            pass
+
+    if not stamped:
+        alembic.command.stamp(alembic.config.Config("conf/alembic.ini"), "head")
+
+
+def includeme(config):  # pragma: no cover
+    engine = create_engine(config.registry.settings["database_url"])
+
+    def db_session(request):
+        """Return the SQLAlchemy session for the given request."""
+        session = Session(engine, info={"request": request})
+
+        # Register the session with pyramid_tm/transaction so that they'll
+        # create a new DB transaction for each request and close the session at
+        # the end of the request.
+        zope.sqlalchemy.register(session, transaction_manager=request.tm)
+
+        return session
+
+    if config.registry.settings.get("dev"):
+        # Create the database tables if they don't already exist.
+        Base.metadata.create_all(engine)
```

Comment:
This needs to take alembic into account, see the version over LMS:


https://github.com/hypothesis/lms/blob/main/lms/db/__init__.py#L114


In the current version if using alembic to run migrations `make dev` will try to create the tables again and fail to start.

We need to only use `create_all` when the alambic table/row is not there and run the migrations otherwise.

Reply:
I think this also needs to be removed from the web worker process. Currently there's a problem that Via has 4 gunicorn workers in dev and when you run `make dev` they all hit this line of code at the same time and all try to create the new tables at once.

My thought is to have the normal `gunicorn` processes *not* call `create_all()` and to have a separate process that _just_ calls `create_all()` and then exits. I'll add this separate process to `supervisord-dev.conf` alongside the existing `gunicorn` and other processes, so when you run `make dev` this new process will also get run and create the tables (if necessary) and then exit.

We already have a `bin/make_db` script that might be all we need for this new process...

Reply:
> This needs to take alembic into account, see the version over LMS:
> 
> https://github.com/hypothesis/lms/blob/main/lms/db/__init__.py#L114
> 
> In the current version if using alembic to run migrations `make dev` will try to create the tables again and fail to start.
> 
> We need to only use `create_all` when the alambic table/row is not there and run the migrations otherwise.

I don't think this is right. `create_all()` only creates those tables that don't already exist. It ignores tables that do already exist. That's why I can run `make db` repeatedly and it succeeds each time. The version in LMS looks incorrect to me.

Reply:
>  Via has 4 gunicorn workers in dev and when you run make dev they all hit this line of code at the same time 

You are right, I was probably hitting this while testing migrations here.

I still think the version of this over LMS is superior, it allows people that don't care about migrations to just `make db` to get their environment up to date.

The the version here (and the previous version LMS) will fail every time a migration does something that was not adding a new table. Eg adding new column will break the the app util dev either:

- destroy the DB and `make db` again
- manually remember to run migrations.